mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java

```diff
@@ -172,20 +174,23 @@ private Mailbox addAcl(MailboxACL acl, Mailbox mailbox) {
         String fixedNamespace = query.getFixedNamespace();
         Username fixedUser = query.getFixedUser();
 
-        return listPaths(fixedNamespace, fixedUser)
-            .filter(idAndPath -> query.isPathMatch(idAndPath.getMailboxPath()))
-            .distinct(CassandraIdAndPath::getMailboxPath)
-            .concatMap(this::retrieveMailbox);
+        return listMailboxes(fixedNamespace, fixedUser)
+            .filter(mailbox -> query.isPathMatch(mailbox.generateAssociatedPath()))
+            .distinct(Mailbox::generateAssociatedPath);
     }
 
-    private Flux<CassandraIdAndPath> listPaths(String fixedNamespace, Username fixedUser) {
-        return needMailboxPathV1Support()
+    private Flux<Mailbox> listMailboxes(String fixedNamespace, Username fixedUser) {
+        return needMailboxPathPreviousVersionsSupport()
             .flatMapMany(needSupport -> {
                 if (needSupport) {
-                    return Flux.concat(mailboxPathV2DAO.listUserMailboxes(fixedNamespace, fixedUser),
-                        mailboxPathDAO.listUserMailboxes(fixedNamespace, fixedUser));
+                    return Flux.concat(
+                        mailboxPathV3DAO.listUserMailboxes(fixedNamespace, fixedUser),
+                        Flux.concat(
+                                mailboxPathV2DAO.listUserMailboxes(fixedNamespace, fixedUser),
+                                mailboxPathDAO.listUserMailboxes(fixedNamespace, fixedUser))
+                            .flatMap(this::retrieveMailbox));
```

Comment:
Question (I'm not sure): why a concat in a concat? Why not align all those operations in one single concat (you can put more than 2 arguments), like previously?

Reply:
Because v1 and v2 needs to retrieveMailbox and not v3, I don't want to duplicate the retrieveMailbox step.